[needs-docs] New layer panel action 'Move Out of Group', replaces 'Make Top Level' (deprecated). Small modification of the actions behavior

MieWinstrup opened the pull request:
## Description
The code in this PR changes the name of the layer panel action 'Move to Top Level' to 'Ungroup'.
This new name, 'Ungroup', is more informative of the action and avoids confusion with the action 'Move to Top' (added in version 3.2). 

Furthermore, the action is slightly changed: Now the selected layers in a group are placed just above this group. This is more intuitive and it resembles the behavior of the 'Group Selected'-action. Before the 'Move to Top level'-action moved the selected layers to the bottom of the layer panel.

Unit test of the changed action is added. 

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

MieWinstrup committed:
New action 'Move Out of Group' that moves layer(s) out of group(s). Deprecating the action 'Move to top level'

MieWinstrup committed:
Added test of the action 'Move Out of Group' in the layer panel

MieWinstrup commented:
I don't know if I should tag this PR with [needs-docs] - the userface of QGIS is changed slightly with the name change and the action it self is changed a bit.

DelazJ commented:
@MieWinstrup yes you should add the tag. This will need to be updated in docs. Thanks.

nyalldawson commented on src/gui/layertree/qgslayertreeviewdefaultactions.h:
```diff
@@ -66,7 +66,11 @@ class GUI_EXPORT QgsLayerTreeViewDefaultActions : public QObject
+     * \see ungroup()
+     * \since QGIS 3.2
+     */
+    QAction *actionUngroup( QObject *parent = nullptr ) SIP_FACTORY;
```
We can't change these names in the API, because they are part of the stable API now.

There's also the risk that someone has code which relies on the old behaviour of the actionMakeTopLevel call.

What I'd suggest here is adding ungroup as a new method/action, and the marking makeTopLevel as deprecated.

The old action could exist, but not actually be inserted into the menu.

Then existing code will work without issue and we can drop the deprecated methods at 4.0

MieWinstrup replied:
Sounds like a good idea. I will make the changes.

nyalldawson reviewed (commented):
Code looks good to me, with exception of the API break. @nirvn @delazj are you ok with the new behaviour? (I'm +1 to it)

DelazJ commented:
That said how about using "Move (out?) from group" label and instead keep "Ungroup" label for a feature that will explode a group (remove the group item but keep its layers and sub-groups)?

MieWinstrup commented:
I have changed the name to 'Move Out of Group', added this as a new action, and marked the action 'Make Top Level' as deprecated. Is this the correct way? I get warnings when compiling.

MieWinstrup commented:
@nyalldawson - This is the warning I get when I compile. Is this the intended way for deprecation to work?
![qgis-deprecatedwarning](https://user-images.githubusercontent.com/16276034/39446017-55c4c946-4cbd-11e8-9402-b5c0b9028e7d.png)

nyalldawson commented:
Actually it's on to call the deprecated member here, because it's being called from a method which is deprecated itself. (I.e. they'll both be removed at the same time)

What you need to do to avoid the warnings is wrap the call to the deprecated member in:


    Q_NOWARN_DEPRECATED_PUSH

    ...

    Q_NOWARN_DEPRECATED_POP

nyalldawson commented:
Great work!
